src/base/associative_cache.hh

```diff
@@ -0,0 +1,347 @@
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __CACHE_LIBRARY_HH__
+#define __CACHE_LIBRARY_HH__
+
+#include <vector>
+
+#include "base/cache_entry.hh"
+#include "base/intmath.hh"
+#include "mem/cache/replacement_policies/base.hh"
+#include "mem/cache/replacement_policies/replaceable_entry.hh"
+#include "mem/cache/tags/indexing_policies/base.hh"
+#include "types.hh"
+
+namespace gem5
+{
+
+/** Cache Library.
+ *
+ */
+template <typename Entry>
+class AssociativeCache
+{
+    static_assert(std::is_base_of_v<CacheEntry, Entry>,
+                  "Entry should be derived from CacheEntry");
+  public:
+    typedef replacement_policy::Base BaseReplacementPolicy;
+
+  protected:
+    /** Name of the cache. */
+    std::string cache_name;
+
+    /** The number of entries in the cache. */
+    size_t numEntries;
+
+    /** Associativity of the cache. */
+    size_t _associativity;
+
+    /** The replacement policy of the cache. */
+    BaseReplacementPolicy *replPolicy;
+
+    /** Indexing policy of the cache */
+    BaseIndexingPolicy *indexingPolicy;
+
+    /** Number of sets in the cache. */
+    size_t numSets;
+
+    /** The entries */
+    std::vector<Entry> entries;
+
+  private:
+
+    void initParams(size_t _num_entries, size_t _assoc)
+    {
+        fatal_if(!isPowerOf2(_num_entries), "The number of entries of an "
```

Comment:
There's a lot of code in this file. I suggest splitting out into a .cc and .hh file for maintainability.

Reply:
I tried that approach and it was a bit messy due to having to instantiate the template separately for each use case: https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl

I'd much rather prefer this way, but if there is any way to avoid errors  please let me know. Here is the WIP patch doing this (does not compile, has errors): https://github.com/pranith/gem5/commit/143aed0899c4401e3c7bd31d4c9712d0d998fd25

Reply:
Oh, right. I missed that this was a template class. please ignore this comment :)